src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -949,6 +967,26 @@ - (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSSt
                     MTR_LOG_INFO("%@ report %@ value filtered - same values as cache", self, attributePath);
                 }
             }
+
+            // If General Diagnostics UpTime attribute, save the likely better estimated start time. At OnSubscriptionEstablished
+            // time, this estimate will be used if it's more accurate. Note that this needs to be done because the current SDK
+            // implementation reports events after attributes, and the system startup event resets the estimated start time.
```

Comment:
But this means we can't use the UpTime value for determining likely times for events we read during the priming read...

I think we should stop resetting the estimated start time when we see the startup event: it should just do the normal "it can go backward from where it is now, or set it if not set yet" thing.  Since we reset `_estimatedStartTime` when subscription drops, that should work, no?

Reply:
The problem here is that:
A) We have to respect StartUp event means the device has rebooted, and clear the current estimate.
B) In current implementations events are reported after attributes, and so if we don't save the UpTime attribute, StartUp would render the UpTime estimate moot.

The idea is that the estimated start time may not be accurate until we are more sure.

I suppose we could also take the saved estimate at StartUp event time, if it exists. But that is more dependent on the current quirk of events happening after attributes.

Reply:
> We have to respect StartUp event means the device has rebooted, and clear the current estimate.

Do we?  Let's say the device rebooted 20 mins ago and we finally resubscribe.  The StartUp event is going to give us a worse estimate of when it rebooted in this case than the reading the UpTime attribute, no?

Reply:
> Do we? Let's say the device rebooted 20 mins ago and we finally resubscribe. The StartUp event is going to give us a worse estimate of when it rebooted in this case than the reading the UpTime attribute, no?

That is correct. Which is why the UpTime-derived estimate is saved for after StartUp event happens, and used to update the estimate.

We can do this because if UpTime is reported during priming, then we have expectation that it is current and updated since boot.

Reply:
Marking conversation resolved as the latest commit significantly changed logic.